recipes/vit/train.py

```diff
@@ -32,482 +30,299 @@
 from tqdm import tqdm
 
+from checkpoint import load_auto_resume_checkpoint, save_auto_resumable_checkpoint
+from distributed import initialize_distributed
 from imagenet_dataset import ImageNetDataset, infinite_dataloader
 from imagenet_utils import transforms_imagenet_eval, transforms_imagenet_train
-from vit import VisionTransformer
+from vit import build_vit_model
 
 
 _logger = logging.getLogger(__name__)
 
 
-def load_checkpoint(model, optimizer, checkpoint_path):
-    """Load a Torch DCP checkpoint from checkpoint_path into model and optimizer.
-
-    Docs: https://docs.pytorch.org/docs/stable/distributed.checkpoint.html
-    """
-    # Load model and optimizer checkpoints.
-    state_dict = {
-        "model": model.state_dict(),
-        "optimizer": optimizer.state_dict(),
-    }
-    torch.distributed.checkpoint.load(state_dict, checkpoint_id=checkpoint_path)
-    model.load_state_dict(state_dict["model"])
-    optimizer.load_state_dict(state_dict["optimizer"])
-
-
-def load_auto_resume_checkpoint(cfg, model, optimizer):
-    """Auto-resume training from the latest checkpoint.
-
-    Checkpoint directories should adhere to the simple format: step_<step_idx>_loss_<loss_value>
-    If cfg.training.checkpoint.resume_from_metric is '+' or '-', then the loss_value is utilized
-    for determining the optimal checkpoint to resume from. Otherwise, the latest checkpoint by
-    modification time is chosen for resumption.
-
-    Args:
-        cfg: Hydra config.
-        model: Model to load checkpoints into.
-        optimizer: Optimizer to load checkpoints into.
-
-    Returns:
-        The latest step index to resume from.
-    """
-    # Auto-Resume: Load latest model and optimizer checkpoints.
-    latest_step_idx = 0
-    if cfg.training.checkpoint.path and Path(cfg.training.checkpoint.path).exists():
-        # Get latest checkpoint sub-directory, which should ONLY contain Torch DCP checkpoint sub-directories.
-        subdirs = [x.absolute() for x in Path(cfg.training.checkpoint.path).iterdir() if x.is_dir()]
-        if len(subdirs) > 0:
-            # We expect a checkpoint named as: step_<step_idx>_loss_<loss_value>.
-            # Get the latest step, the directory with the most recent modification time.
-            opt_metric_coeff = 1 if cfg.training.checkpoint.resume_from_metric == "+" else -1
-            latest_subdir = max(
-                subdirs,
-                key=lambda x: (
-                    opt_metric_coeff * float(x.name.split("_")[3])
-                    if cfg.training.checkpoint.resume_from_metric
-                    else 0,
-                    x.stat().st_mtime,
-                ),
+@hydra.main(version_base="1.2", config_path="config", config_name="vit_base_patch16_224")
+def main(cfg) -> None:
+    """Train a ViT model on ImageNet using Megatron-FSDP and TransformerEngine (TE)."""
+
+    # Initialize distributed environment.
+    with initialize_distributed(cfg) as device_mesh:
```

Comment:
imo would be clearer as `initialize_distributed(**cfg.distributed)`; where `initialize_distributed` then took standard, typed kwargs (dp, cp, etc.), rather than giving it the entire object